Misc.Fixes 2

alyst opened the pull request:
Second round of cherry-picking from #193.

This time there are:
* some linear algebra tweaks in ML
* indexing optimizations, esp. in FABIN3. Some helper functions are not actually used in the current code, so maybe they just could be removed.
* some test enhancements: essentially instead of doing `AeqB = A == B; @test AeqB` we do `@test A == B` directly, which provides better diagnostic when the test fails

These commits do not affect the API (except that `n_man` and `n_obs` are now restricted to `Int` -- in all of their usages in the current code they are converted back to Int from Float64).

Maximilian-Stefan-Ernst commented on src/additional_functions/helper.jl:
```diff
@@ -1,10 +1,14 @@
-function neumann_series(mat::SparseMatrixCSC)
+# Neumann series representation of (I - mat)⁻¹
+function neumann_series(mat::SparseMatrixCSC; maxn::Integer = size(mat, 1))
```
Nice change - it is a fact that if the Neumann series converges in a finite number of steps, it has to be less than `size(mat, 1)`, right?

alyst replied:
My intuition is that if the matrix has lower triangular form, it should have zeroes on the main diagonal (to converge in finite steps).
Then at each multiplication non-zeroes should move away from the main diagonal, and disappear in at most `size(mat, 1)` steps.

Maximilian-Stefan-Ernst replied:
Nice. A fun fact I remembered: since the `A` matrix contains the directed effects, the number of iterations of the Neumann series corresponds to the length of the longest directed path in the model 😄 So if your model contains directed loops (is not a DAG), the series won't converge in finite steps.

Maximilian-Stefan-Ernst commented on src/additional_functions/helper.jl:
```diff
@@ -1,10 +1,14 @@
-function neumann_series(mat::SparseMatrixCSC)
+# Neumann series representation of (I - mat)⁻¹
+function neumann_series(mat::SparseMatrixCSC; maxn::Integer = size(mat, 1))
```
```suggestion
function neumann_series(mat::SparseMatrixCSC; maxiter::Integer = size(mat, 1))
```

Maximilian-Stefan-Ernst commented on src/additional_functions/helper.jl:
```diff
@@ -1,10 +1,14 @@
 
+    n = 1
     while nnz(next_term) != 0
+        (n <= maxn) || error("Neumann series did not converge in $maxn steps")
```
```suggestion
        (n <= maxiter) || error("Neumann series did not converge in $maxn steps")
```

alyst replied:
```suggestion
        (n <= maxiter) || error("Neumann series did not converge in $maxiter steps")
```

Maximilian-Stefan-Ernst commented on src/additional_functions/helper.jl:
```diff
@@ -37,11 +41,8 @@ function get_observed(rowind, data, semobserved; args = (), kwargs = NamedTuple(
-    for i in 1:size(mat, 2)
-        @views means[i] = mean(skipmissing(mat[:, i]))
-    end
+function skipmissing_mean(mat::AbstractMatrix)
+    means = [mean(skipmissing(coldata)) for coldata in eachcol(mat)]
     return means
 end
```
If you want we could even do
```suggestion
skipmissing_mean(mat::AbstractMatrix) = 
    [mean(skipmissing(coldata)) for coldata in eachcol(mat)]
```

Maximilian-Stefan-Ernst commented on src/additional_functions/parameters.jl:
```diff
@@ -111,97 +115,23 @@ function fill_matrix!(
-                ),
-            )
-            return nothing
+# range of parameters that are referenced in the matrix
+function param_range(mtx_indices::AbstractArrayParamsMap)
+    first_i = findfirst(!isempty, mtx_indices)
+    last_i = findlast(!isempty, mtx_indices)
+
+    if !isnothing(first_i) && !isnothing(last_i)
+        for i in first_i:last_i
+            if isempty(mtx_indices[i])
+                # TODO show which parameter is missing in which matrix
+                throw(
+                    ErrorException(
+                        "Your parameter vector is not partitioned into directed and undirected effects",
+                    ),
+                )
+            end
```
Very nice change. It seems like this is a function that is used nowhere...

alyst replied:
Yes, there are quite a few helper functions that don't seem to be utilized atm (coverage provides good uptodate overview). I think once most of the proposed changes land, it would be easier to figure out what to do with it (e.g. if they could be used in any future).

Maximilian-Stefan-Ernst commented on src/observed/covariance.jl:
```diff
@@ -39,11 +39,11 @@ use this if you are sure your covariance matrix is in the right format.
 - `spec_colnames::Vector{Symbol} = nothing`: overwrites column names of the specification object
 """
-struct SemObservedCovariance{B, C, D, O} <: SemObserved
+struct SemObservedCovariance{B, C} <: SemObserved
```
I'm wondering what best practices is here - I think the old version should not have downsides in terms of performance; restricting the types of these two fields is maybe best done with:
```suggestion
struct SemObservedCovariance{B, C, D <: Int, O <: Int} <: SemObserved
```
I'm not sure when/if annotating the types of specific fields vs. having a parametric type is preferred, maybe you have some thoughts on that?

alyst replied:
As far as I know, type parameters add some stress on Julia compiler, so if they could be avoided (i.e. if the type is fixed), they should be avoided.
Plus, it makes the code less readable.
In this particular case, it is quite reasonable to fix the number of observations and the number of variables to `Int`.
I don't see usecases, where anything else would be required.

alyst replied:
Also, `D <: Int` is, sort of, no-op as `Int` is a concrete type, i.e. it just tells that `D === Int`.
`D <: Integer` allows flexibility, since `Integer` is an abstract type, but, as I wrote in the previous comment, I don't see the benefit of this flexibility here.

alyst replied:
BTW, here's the [performance tips about type declarations](https://docs.julialang.org/en/v1/manual/performance-tips/#Type-declarations) from Julia manual

Maximilian-Stefan-Ernst replied:
I see, then let's do it like you suggested. In the manual it only says if you have abstract field types it's better to use parametric Types; but here the field types are concrete (so I guess that makes it okay).

Maximilian-Stefan-Ernst commented on test/unit_tests/multithreading.jl:
```diff
@@ -2,6 +2,6 @@ using Test
 if haskey(ENV, "JULIA_ON_CI")
     @testset "multithreading_enabled" begin
-        @test Threads.nthreads() == 8
+        @test Threads.nthreads() >= 8
```
Is there a specific reason for this? This test should be only run on CI, where I set the number of threads to 8.

Maximilian-Stefan-Ernst commented on src/additional_functions/start_val/start_fabin3.jl:
```diff
@@ -86,101 +80,101 @@ function start_fabin3(ram_matrices::RAMMatrices, Σ, μ)
     # ind_Λ = findall([is_in_Λ(ind_vec, F_ind) for ind_vec in A_ind_c])
 
-    for i in findall(.!(1:n_nod .∈ [F_ind]))
+    function calculate_lambda(
```
Is there a reason to have this function definition inside the other function?

alyst replied:
It could be moved outside (just Σ has to be added as another input parameter), but it is only used from start_fabin3 so far, and to me it looked like both its purpose and reference/indicator terminology are FABIN3-specifc.

Maximilian-Stefan-Ernst commented on src/additional_functions/start_val/start_fabin3.jl:
```diff
@@ -86,101 +80,101 @@ function start_fabin3(ram_matrices::RAMMatrices, Σ, μ)
-            if size(reference, 1) > 1
-                @warn "You have more than 1 scaling indicator"
-                reference = reference[1]
+        if length(reference) > 0
+            if length(reference) > 1
+                if isempty(indicator2parampos) # don't warn if entire column is fixed
+                    @warn "You have more than 1 scaling indicator for $(ram_matrices.colnames[i])"
+                end
+                ref = reference[1]
```
```suggestion
        if length(reference) > 0
            ref = reference[1]
            if (length(reference) > 1) & isempty(indicator2parampos) # don't warn if entire column is fixed
                    @warn "You have more than 1 scaling indicator for $(ram_matrices.colnames[i])"
            end
```

Maximilian-Stefan-Ernst replied:
Sorry, GitHub does not let me make this suggestion sensible, but I wanted to also delete the following `else` case.

alyst commented on src/imply/RAM/generic.jl:
```diff
@@ -232,7 +224,12 @@ function objective!(imply::RAM, parameters, model, has_meanstructure::Val{T}) wh
     )
 
-    imply.I_A .= I - imply.A
+    @inbounds for (j, I_Aj, Aj) in
+                  zip(axes(imply.A, 2), eachcol(imply.I_A), eachcol(imply.A))
+        for i in axes(imply.A, 1)
+            I_Aj[i] = ifelse(i == j, 1, 0) - Aj[i]
+        end
+    end
```
Actually, this one is simpler and might be faster:
```suggestion
    axpby!(-1, imply.A, 0, imply.I_A)
    @inbounds imply.I_A[diagind(imply.I_A)] .+= 1
```
It is also quite explicit to deserve a function.

For a function, I'd generalize it to expressions like `C := a I + b B + c C` or at least `C := a I + b B`

Maximilian-Stefan-Ernst replied:
Benchmarkingit, `axpby!` is a bit slower, but this one is faster:
```suggestion
    @. I_A = -A
    @view(I_A[diagind(I_A)]) .+= 1
```
and also easy enough that it does not need a function.

Maximilian-Stefan-Ernst reviewed (changes requested):
Very nice PR with many improvements. There's only one open comment I think and then we can merge.

alyst commented:
Thanks again for very fast review!
I've updated I-A calculation, integrated the review changes into original commits and rebased the PR.

The next round of changes may substantially change some internal structs like RAMMatrices. So potentially you can release the current fixes as the new patch version.

Maximilian-Stefan-Ernst commented:
Ah thanks for re-fixing the I-A computation^^
I will release this as a patch.
